contracts/upgradeables/chips/Chips.sol

```diff
@@ -0,0 +1,246 @@
+import {
+    SafeERC20,
+    IERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {
+    IERC20Metadata
+} from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import {
+    AccessControlUpgradeable
+} from "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import {
+    Initializable
+} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {
+    PausableUpgradeable
+} from "@openzeppelin/contracts-upgradeable/utils/PausableUpgradeable.sol";
+import { Strings } from "@openzeppelin/contracts/utils/Strings.sol";
+
+import {
+    ReentrancyGuardUpgradeable
+} from "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+import {
+    UUPSUpgradeable
+} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+contract Chips is
+    IChips,
+    Initializable,
+    AccessControlUpgradeable,
+    PausableUpgradeable,
+    ReentrancyGuardUpgradeable,
+    UUPSUpgradeable
+{
+    using SafeERC20 for IERC20;
+
+    event ChipsUpdate(address indexed from, address indexed to, uint256 value);
+    event Deposit(address indexed user, uint256 amount);
+    event Withdraw(address indexed user, uint256 amount);
+    event AdminDeposit(
+        address indexed admin,
+        address[] indexed users,
+        uint256[] amounts
+    );
+    event WithdrawAllAdmin(address indexed admin, address[] indexed users);
+    event RetrieveBuyIn(
+        address indexed game,
+        address indexed user,
+        uint256 amount
+    );
+    event DistributeChips(
+        address indexed game,
+        address[] indexed users,
+        uint256[] amounts
+    );
+
+    error ChipInsufficientBalance(
+        address from,
+        uint256 fromBalance,
+        uint256 value
+    );
+
+    error ArrayLengthMismatch();
+
+    bytes32 public constant GAME_ROLE = keccak256("GAME_ROLE");
+    bytes32 public constant MANAGER_ROLE = keccak256("MANAGER_ROLE");
+
+    IERC20 public token;
+    mapping(address => uint256) public _balances;
+    uint256 public _totalSupply;
+    uint8 private _decimals;
+
+    // @dev Initializes the contract
+    // @param _token The address of the token to use for the chips
+    // @param _isPaused Whether the contract is paused
+    function initialize(address _token, bool _isPaused) public initializer {
+        __ReentrancyGuard_init();
+        __Pausable_init();
+        __AccessControl_init();
+        __UUPSUpgradeable_init();
+
+        _grantRole(MANAGER_ROLE, msg.sender);
+        _setRoleAdmin(MANAGER_ROLE, MANAGER_ROLE);
+        _setRoleAdmin(GAME_ROLE, MANAGER_ROLE);
+
+        token = IERC20(_token);
+        _decimals = IERC20Metadata(_token).decimals();
+
+        if (_isPaused) _pause();
+    }
+
+    // @dev Deposits the chips to the user
+    // @param amount The amount of chips to deposit
+    function deposit(uint256 amount) external whenNotPaused nonReentrant {

```

Comment:
Please remember to extend from the whitelist upgradeable contract to verify on-chain what the wallet server does. I only want to allow deposits from our portal